Kerberos (Rebased #3417)

akallabeth opened the pull request:
Rebased #3417 to current master

HenryJacques commented:
If I may, I think it will also be useful to add the following three cases in `transport.c` in the `switch (ret)`:

KRB5_KDC_UNREACH: Cannot contact any KDC for requested realm 
KRB5KDC_ERR_CLIENT_REVOKED: Clients credentials have been revoked 
KRB5KDC_ERR_POLICY: KDC policy rejects request

(of course you can add as many Kerberos errors as you want but with these three, you cover 90% of the errors. See https://tools.ietf.org/html/rfc1510#page-84)

akallabeth commented:
@HenryJacques Ok, will have a look tomorrow.

mfleisz commented on libfreerdp/core/transport.c:
```diff
@@ -59,17 +59,187 @@
+	krb5_ccache ccache;
+	krb5_init_creds_context ctx = NULL;
+	memset(address, 0, sizeof(address));
+	strcpy(address, username);
```
_snprintf(... , "%s@%s", username, domain) would be safer (checks for buffer size) and we can also get rid of the memset

hardening commented on libfreerdp/core/transport.c:
```diff
@@ -59,17 +59,187 @@
+	if (!g_KRB5Module)
+	{
+		WLog_ERR(TAG, "Failed to load krb5 module\n");
+	}
```
I guess we should exit in this case, not really sure if GetProcAddress(NULL) is safe...

hardening commented on libfreerdp/core/transport.c:
```diff
@@ -59,17 +59,187 @@
+	g_KRB5.pKrb5_init_creds_get 		= (void*) GetProcAddress(g_KRB5Module, "krb5_init_creds_get");
+	g_KRB5.pKrb5_init_creds_free 		= (void*) GetProcAddress(g_KRB5Module, "krb5_init_creds_free");
+	g_KRB5.pKrb5_free_context 		= (void*) GetProcAddress(g_KRB5Module, "krb5_free_context");
+
```
All these should be checked against NULL or we may run into troubles some lines later...

akallabeth commented:
@mfleisz @hardening Ok, reworked the kerberos library stuff, directly linking it in now. No more ```LoadLibrary``` et al, just plain CMake detection. Also added the error handlers @HenryJacques suggested.

hardening commented on libfreerdp/core/transport.c:
```diff
@@ -59,17 +59,130 @@
+	if (ret != KRB5KDC_ERR_NONE)
+		goto out;
+	/* Create a context for acquiring initial credentials */
+	else if ((ret = krb5_init_creds_init(context, principal, NULL, NULL, 0, NULL, &ctx)) != 0)
```
the else is useless

hardening commented on libfreerdp/core/transport.c:
```diff
@@ -59,17 +59,130 @@
+		goto out;
+	}
+	/* Set a password for acquiring initial credentials */
+	else if ((ret = krb5_init_creds_set_password(context, ctx, passwd)) != 0)
```
the same here

hardening commented on libfreerdp/core/transport.c:
```diff
@@ -59,17 +59,130 @@
+				freerdp_set_last_error(transport->context, FREERDP_ERROR_AUTHENTICATION_FAILED);
+			}
+		}
+		else if (ret == 592)
```
Just noticed that, but it would be neat to have a define or at least a comment for that mysterious 592 value..

HenryJacques replied:
Looking at https://linux.die.net/man/3/ssl_set_info_callback the `ret` value can be interpreted with `ssl_alert_type_string`. This function return F (as Fatal)  

`case SSL3_AL_FATAL:
        return "F";`

HenryJacques commented on libfreerdp/core/transport.c:
```diff
@@ -59,17 +59,130 @@
+		case KRB5KDC_ERR_NONE:
+			break;
+
+		case KRB5_KDC_UNREACH:
```
KRB5_KDC_UNREACH shouldn't be in this case. In my opinion, it should return a specific message in WLog_WARN indicating it cannot contact the Kerberos server (you can keep the same error FREERDP_ERROR_AUTHENTICATION_FAILED)

HenryJacques commented on libfreerdp/core/transport.c:
```diff
@@ -59,17 +59,130 @@
+			break;
+
+		case KRB5KDC_ERR_CLIENT_REVOKED:
+			ret = FREERDP_ERROR_CONNECT_PASSWORD_CERTAINLY_EXPIRED;
```
I don't agree. FREERDP_ERROR_CONNECT_PASSWORD_CERTAINLY_EXPIRED is only used for non-domain workstations. In this case the Kerberos server was reached. We should define a new error code, for example FREERDP_ERROR_CONNECT_CLIENT_REVOKED

HenryJacques commented on libfreerdp/core/transport.c:
```diff
@@ -59,17 +59,130 @@
+			break;
+
+		case KRB5KDC_ERR_POLICY:
+			ret = FREERDP_ERROR_INSUFFICIENT_PRIVILEGES;
```
In my opinion, we could add a WLog_WARN message indicating the reason of this error. During my tests I had this error only when the account was forced to use smartcard logon (and not login/password)

HenryJacques commented:
@akallabeth I did put 3 single comments in the PR - `in transport.c` (they don't appear in the conversation as they are not reviews I think)

akallabeth commented:
@mfleisz @hardening @HenryJacques Ok, fixed all your comments and modified the logger to use the faster preloaded one. Don't be fooled by the failing compile, we need to install ```libkrb5-dev``` on the machines before it will go through.

mfleisz commented on libfreerdp/core/transport.c:
```diff
@@ -59,17 +59,146 @@
+	if (ctx)
+		krb5_init_creds_free(context, ctx);
+
+	krb5_free_context(context);
```
if krb5_init_context fails we potentially free an invalid context here

akallabeth commented:
@freerdp-bot test

hardening commented:
Would be nice to drop the useless else, anyway it's not a show stopper.
